FirebaseFunctionsSwift/Tests/Unit/SerializerTests.swift

```diff
@@ -63,186 +63,199 @@ class SerializerTests: XCTestCase {
+      return
+    }
+    XCTFail()
+  }
+
+  func testEncodeUnsignedLong() throws {
+    let serializer = FUNSerializer()
+    let typeString = "type.googleapis.com/google.protobuf.UInt64Value"
+    let highULong = NSNumber(value: 18_446_744_073_709_551_607 as UInt64)
+    let expected = ["@type": typeString, "value": "18446744073709551607"]
+    let encoded = try serializer.encode(highULong) as? [String: String]
+    XCTAssertEqual(encoded, expected)
+  }
+
+  func testDecodeUnsignedLong() throws {
+    let serializer = FUNSerializer()
+    let highULong = NSNumber(value: 18_446_744_073_709_551_607 as UInt64)
+    XCTAssertEqual(highULong, try serializer.decode(highULong) as? NSNumber)
+  }
+
+  func testDecodeUnsignedLongFromDictionary() throws {
+    let serializer = FUNSerializer()
+    let typeString = "type.googleapis.com/google.protobuf.UInt64Value"
+    let highULong = NSNumber(value: 18_446_744_073_709_551_607 as UInt64)
+    let coded = ["@type": typeString, "value": "18446744073709551607"]
+    let decoded = try serializer.decode(coded) as? NSNumber
+    XCTAssertEqual(highULong, decoded)
+    // A naive implementation might convert a number to a double and think that's close enough.
+    // We need to make sure it's an unsigned long long for accuracy.
+    XCTAssertEqual(decoded?.objCType[0], CChar("Q".utf8.first!))
+  }
+
+  func testDecodeUnsignedLongFromDictionaryOverflow() throws {
+    let serializer = FUNSerializer()
+    let typeString = "type.googleapis.com/google.protobuf.UInt64Value"
+    let tooHighVal = "18446744073709551616"
+    let coded = ["@type": typeString, "value": tooHighVal]
+    do {
+      _ = try serializer.decode(coded) as? NSNumber
+    } catch let SerializerError.invalidValueForType(value, type) {
+      XCTAssertEqual(value, tooHighVal)
+      XCTAssertEqual(type, typeString)
+      return
+    }
+    XCTFail()
+  }
+
+  func testEncodeDouble() throws {
+    let serializer = FUNSerializer()
+    let myDouble = NSNumber(value: 1.2 as Double)
+    XCTAssertEqual(myDouble, try serializer.encode(myDouble) as? NSNumber)
+  }
+
+  func testDecodeDouble() throws {
+    let serializer = FUNSerializer()
+    let myDouble = NSNumber(value: 1.2 as Double)
+    XCTAssertEqual(myDouble, try serializer.decode(myDouble) as? NSNumber)
+  }
+
+  func testEncodeBool() throws {
+    let serializer = FUNSerializer()
+    XCTAssertEqual(true, try serializer.encode(true) as? NSNumber)
+  }
+
+  func testDecodeBool() throws {
+    let serializer = FUNSerializer()
+    XCTAssertEqual(true, try serializer.decode(true) as? NSNumber)
+  }
+
+  func testEncodeString() throws {
+    let serializer = FUNSerializer()
+    XCTAssertEqual("hello", try serializer.encode("hello") as? String)
+  }
+
+  func testDecodeString() throws {
+    let serializer = FUNSerializer()
+    XCTAssertEqual("good-bye", try serializer.decode("good-bye") as? String)
+  }
+
+  // TODO: Should we add support for Array as well as NSArray?
```

Comment:
I think it implicitly works right now as it bridges to the ObjC implementation, but good question on whether or not it will do that now. Ideally yes if we can, but let's see what it looks like from the API surface (if you need to cast as NSArray or it implicitly does so)

Reply:
Yeah, I think the API surface might be fine - since we didn't notice this with the integration tests.